docs/source/building/variants.rst

```diff
@@ -0,0 +1,676 @@
+not independent of input versions, don't use this key!
+
+Any pinning done in the run requirements will affect the hash, and thus builds will
+be done for each variant in the matrix.
+
+
+CONDA_* variables and command line arguments to conda-build
+-----------------------------------------------------------
+
+To ensure legacy consistency, environment variables such as CONDA_PY behave as
+they always have, and they clobber all variants set in files or passed to the
+API.
+
+The full list of respected environment variables are:
+
+* CONDA_PY
+* CONDA_NPY
+* CONDA_R
+* CONDA_PERL
+* CONDA_LUA
+
+Legacy CLI flags are also still available. These are sticking around for their
+usefulness in one-off jobs.
+
+* --python
+* --numpy
+* --R
+* --perl
+* --lua
+
+
+Aggregation of multiple variants
+--------------------------------
+
+The matrix of all variants is first consolidated from several dicts of lists
+into a single dict of lists, and then transformed in a list of dicts (via the
+Cartesian product of lists), where each value is a single string from the list
+of potential values.
+
+For example, general input for ``variants`` could be something like:
+
+.. code-block:: python
+
+    a = {'python': ['2.7', '3.5'], 'numpy': ['1.10', '1.11']}
+    # values can be strings or lists.  Strings are converted to one-element lists internally.
+    b = {'python': ['3.4', '3.5'], 'numpy': '1.11'}
+
+
+Here, let's say ``b`` is found after ``a``, and thus has priority over ``a``. Merging these
+two variants yields:
+
+.. code-block:: python
+
+    merged = {'python': ['3.4', '3.5'], 'numpy': ['1.11']}
+
+
+``b``'s values for ``python`` have clobbered ``a``'s. From here, we compute the
+Cartesian product of all input variables. The end result is a collection of
+dicts, each with a string for each value. Output would be something like:
+
+.. code-block:: python
+
+    variants = [{'python': '3.4', 'numpy': '1.11'}, {'python': '3.5', 'numpy': '1.11'}]
+
+
+and conda-build would loop over these variants where appropriate (building,
+outputting package output names, etc.)
+
+If ``numpy`` had had two values instead of one, we'd end up with *four* output
+variants: 2 variants for ``python``, *times* two variants for ``numpy``:
+
+.. code-block:: python
+
+    variants = [{'python': '3.4', 'numpy': '1.11'}, {'python': '3.5', 'numpy': '1.11'},
+                {'python': '3.4', 'numpy': '1.10'}, {'python': '3.5', 'numpy': '1.10'}]
+
+
+Bootstrapping pins based on an existing environment
+---------------------------------------------------
+
+
+To establish your initial variant, you may point at an existing conda
+environment. Conda-build will examine the contents of that environment and pin
+to the exact requirements that make up that environment.
+
+.. code-block:: shell
+
+   conda build --bootstrap name_of_env
+
+
+You may specify either environment name (and depend on conda's environment
+lookup) or filesystem path to the environment.
+
+
+Extended keys
+-------------
```

Comment:
And example of extended keys would be great.  The concept is still a little nebulous for me.  What are the use cases?